### ISSUE TITLE
Populate context.options with the default values when no options are present

### ISSUE THREAD

**Issue by GMartigny** [eslint/eslint#13903]
**The version of ESLint you are using.**

	7.14.0

**The problem you want to solve.**

When writing a plugin, I want to specify some default value for the rules. However, when the end-user don't specify any option, nothing is passed to the rule context.

This kind of kill the point of `default` values in the schema.

**Your take on the correct solution to problem.**

Apparently, `AJV` [fill the schema with the default value in place](https://github.com/ajv-validator/ajv/blob/master/docs/validation.md#assigning-defaults). It means that is no object is passed down, it can't fill it.

I think, passing an empty object if no options are specified could solve the issue.

**Are you willing to submit a pull request to implement this change?**

Yes

I already made a [minimal reproduction repo](https://github.com/GMartigny/eslint-no-default).

**Comment by mdjermanovic** [comment]
Hi @GMartigny, thanks for the issue!

This would be problematic for rules with `"oneOf"` in the schema, like [array-bracket-newline](https://github.com/eslint/eslint/blob/110cf962d05625a8a1bf7b5f4ec2194db150eb32/lib/rules/array-bracket-newline.js#L29). In such cases we can't know what's intended default for a rule (i.e., how it handles the absence of configured options), so passing an empty object to AJV could break the behavior.

There are other problems, e.g., what to do with nested objects and how to handle ["required"](https://github.com/eslint/eslint/blob/110cf962d05625a8a1bf7b5f4ec2194db150eb32/lib/rules/no-multiple-empty-lines.js#L42).

We could maybe do this just for rules with a simple schema, but that would be inconsistent and thus confusing.

**Comment by GMartigny** [comment]
I only open this issue for schema with a `default` property. You're right, there is a lot of corner and special cases.

I might miss it, but I don't any issue with `"required"` or nested object. In my opinion, the rule context options should receive all the `default` entered in the schema.

**Comment by mdjermanovic** [comment]
> I only open this issue for schema with a `default` property. You're right, there is a lot of corner and special cases.

Yes, it seems very complicated to determine if it's safe to pass an empty object that will be populated by defaults, and even more complicated to document and understand in what cases can a rule count on that, so I believe we'll end up with behavior which would be pretty much unpredictable for rule developers.

> I might miss it, but I don't any issue with `"required"` or nested object. In my opinion, the rule context options should receive all the `default` entered in the schema.

Required properties must be specified, so we can't create an object without them (ajv will throw an error, like in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLW11bHRpcGxlLWVtcHR5LWxpbmVzOiBbZXJyb3IsIHt9XSAqLyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6NSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)). That's probably just one of many details that should be considered.

**Comment by futpib** [comment]
Allowing `default` JSON Schema keyword and ignoring it is itself misleading.

Please make a choice wether ESLint should support `default` JSON Schema keyword and then either
* If it should
* * Either make a semver-major change
* * Or make it an opt-in feature
* If it should not
* * Ban the `default` JSON Schema keyword (would be a semver-major change too; ajv [seems to provide such an option already](https://github.com/ajv-validator/ajv/blob/master/docs/strict-mode.md#prohibit-ignored-defaults))

**Comment by mdjermanovic** [comment]
`default` isn't ignored. It's triggered for a property when the provided data matches the schema in a way that its parent object's schema is used. Ajv doesn't create objects, it just fills provided objects with default values for missing properties.

It may seem reasonable to pass an empty object to Ajv (as suggested) for a very simple schema, but it's difficult to generalize this behavior in a predictable way, because schema for a rule can be anything defined by the [JSON Schema](https://json-schema.org/) spec.

For example, what should be the expected `context.options[0]` if the end-user didn't specifiy anything, and rule has the following schema:

```js
schema: [
  {
      type: "object",
      properties: {
          foo: {
              type: "boolean",
              default: true
          },
          bar: {
              type: "string"
          }
      },
      required: ["bar"]
  }
]
```

**Comment by GMartigny** [comment]
> For example, what should be the expected context.options[0] if the end-user didn't specify anything, and rule has the following schema:

Given that `bar` is required and the user didn't specify anything, I would say it raise an error. I just tested and I don't understand why there's a **difference** between "no options" and "empty option". 

> It may seem reasonable to pass an empty object to Ajv (as suggested) for a very simple schema, but it's difficult to generalize

This was a naive and quick test solution. In no way I expected it to be the final solution. The generalization problem is complex given the wide range of valid schema. In the other hand, I don't think `eslint` should support esoteric schema.

**Comment by mdjermanovic** [comment]
> why there's a difference between "no options" and "empty option"

That's how Ajv works, probably for a good reason. Is there a JSON Schema validator that works differently? I think that modifying the schema validation process to a (most likely) uncommon behavior is too complex and out of scope for ESLint.

What does seem to work with Ajv is setting defaults for objects, which then also get validated against the object's schema in the same way, including applying defaults:

```js
const schema = {
  type: "array",
  items: [
      {
          type: "object",
          properties: {
              first: {
                  type: "string",
                  default: "Here I am !"
              },
              second: {
                  type: "string",
                  default: "Me too"
              }
          },
          additionalProperties: false,
          default: {} // <--
      }
  ],
  minItems: 0,
  maxItems: 1
};

const Ajv = require("ajv");

const ajv = new Ajv({ useDefaults: true });

const validate = ajv.compile(schema);

const data = [ /* nothing */ ];

validate(data);

console.log(data); // [ { first: 'Here I am !', second: 'Me too' } ]
```

So it looks like you could add `default: {}` in the rule's schema to enforce creating an object with default values when user doesn't specify anything.

An issue with this approach is that ESLint currently doesn't support `default` in top-level arrays, as noticed in https://github.com/eslint/eslint/issues/11749. We're passing a temporary array ([`Array.isArray(options) ? options.slice(1) : []`](https://github.com/eslint/eslintrc/blob/17468407c1baf05747cb261c91f7f7b7c2a82422/lib/shared/config-validator.js#L136)) to Ajv, so any changes that validation makes directly on the provided array are lost, unlike changes on deep objects. I think we can consider making this enhancement/bug fix.

**Comment by GMartigny** [comment]
> I think we can consider making this enhancement/bug fix.

That would be very nice.
